README.md
# FailToBan

[![Build Status](https://travis-ci.org/jobteaser/fail_to_ban.svg?branch=master)](https://travis-ci.org/jobteaser/fail_to_ban)
[![Code Climate](https://codeclimate.com/repos/58667ffceab18f66d7000836/badges/86fb76a0e71dd832bdea/gpa.svg)](https://codeclimate.com/repos/58667ffceab18f66d7000836/feed)
![Dependencies](https://img.shields.io/badge/dependencies-none-green.svg)

Handle brute force on key backed by Redis.

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'fail_to_ban', git: 'https://github.com/jobteaser/fail_to_ban.git', tag: 'v0.2.1'
```

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install fail_to_ban

## Usage

We provide a default backoff strategy, see: [backoff_strategy.rb](lib/fail_to_ban/strategies/backoff_strategy.rb)

You can make your own custom strategy and inject it, your strategy will be initialized with the following parameters: `key`, `storage`, `config`

The following methods should be implemented:

|Name     |Type   |Description|
|---------|-------|-----------|
|attempt  |symbol |Make a connection attempt, could increment a retry_count for example.|
|blocked? |boolean|Should Tell wether the key should be blocked for another attempt.        |
|reset    |symbol |Should remove the key in storage.                            |
|unlock_at|Time   |Should return the time when the lock will be lifted.                     |

```ruby
$redis = Redis.new

FailToBan.new(
  key: 'hello/42', # Can be any string, should be unique
  storage: $redis, # A key value storage compatible with the redis api
  strategy: Strategies::BackoffStrategy, # Optionnal
  config: { permitted_attempts: 3, backoff_step: 15 } # Optionnal
)
```

### An example with the default `BackoffStrategy`

#### Config parameters

|Parameter         |Type   |Description|Default|
|------------------|-------|-----------|-------|
|permitted_attempts|Integer|The number of permitted attemps before locking the key.|3|
|backoff_step      |Integer|The time the key will be locked for, in seconds, and incrementing after every fail.|15|
|backoff_cap       |Integer|The time after which the unlock_at will stop being incremented. (nil = no limit)|nil|

```ruby

$redis = Redis.new

FailToBan.new(
  key: 'dev@jobteaser.com',
  storage: $redis,
  strategy: Strategies::BackoffStrategy, # By default
  config: { permitted_attempts: 3, backoff_step: 15 }
)

protection.blocked?
# => false

protection.attempt
# => :ok

# Backoff : after 3 failed attempts there is a 15 seconds wait
# If it fails again then it's 30 seconds, then 45,
# In any case, set a +/- 10% jitter on the wait (e.g 14, 28, 47, ...)
protection.protect
# => :blocked

# this method reset blocked key
protection.reset
# => :ok

# this method returns ETA when account
# will be unblocked
protection.unlock_at
# => timestamp

# this method returns the time left before
# the account will be unblocked
protection.unlock_at
# => timestamp

```

## Development

After checking out the repo, run `bin/setup` to install dependencies. Then, run `rake spec` to run the tests. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and tags, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/jobteaser/fail_to_ban. This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [Contributor Covenant](http://contributor-covenant.org) code of conduct.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

### [0.2.1] - 2018-07-10

#### Changed
- Fix error with integer in `unlock_in` method


spec/spec_helper.rb
require 'rspec'

if ENV.fetch('CODECLIMATE_REPO_TOKEN', false)
  require 'simplecov'
  SimpleCov.start
end

$LOAD_PATH.unshift File.expand_path('../../lib', __FILE__)
require 'fail_to_ban'


spec/fail_to_ban_spec.rb
require 'spec_helper'

class NullStrategy
  def initialize(key:, storage:, config: {}); end
end

RSpec.describe FailToBan do
  subject do
    described_class.new(key: 'boring_key', storage: nil, strategy: NullStrategy)
  end

  describe '#attempt' do
    it 'relies on the strategy' do
      expect_any_instance_of(NullStrategy).to receive(:attempt)
      subject.attempt
    end
  end

  describe '#blocked?' do
    it 'relies on the strategy' do
      expect_any_instance_of(NullStrategy).to receive(:blocked?)
      subject.blocked?
    end
  end

  describe '#reset' do
    it 'relies on the strategy' do
      expect_any_instance_of(NullStrategy).to receive(:reset)
      subject.reset
    end
  end

  describe '#unlock_at' do
    it 'relies on the strategy' do
      expect_any_instance_of(NullStrategy).to receive(:unlock_at)
      subject.unlock_at
    end
  end

  describe '#unlock_in' do
    it 'relies on the strategy' do
      allow(Time).to receive(:now).and_return(Time.new('2014', '12', '11').utc)
      expect_any_instance_of(NullStrategy)
        .to receive(:unlock_at)
        .and_return(Time.new('2014', '12', '12').utc.to_i)
      expect(subject.unlock_in).to eq(24 * 60 * 60) # 1 day in seconds
    end
  end

  it 'has a version number' do
    expect(FailToBan::VERSION).not_to be nil
  end
end


spec/fail_to_ban/strategies/backoff_strategy_spec.rb
require 'spec_helper'

class NullStorage

  def initialize(null_data = {})
    @null_data = null_data
  end

  def hset(k, hk, hv)
    @null_data[k] ||= {}
    @null_data[k][hk] = hv.to_s
  rescue NoMethodError
    nil
  end

  def hincrby(k, hk, i)
    @null_data[k] ||= {}
    @null_data[k][hk] = @null_data[k][hk].to_i + i
  end

  def hget(k, hk)
    @null_data[k][hk]
  rescue NoMethodError
    nil
  end

  def expire(k, t)
    nil
  end

  def del(k)
    @null_data.delete(k)
  end

  def flushdb
    @null_data.clear
  end

end

RSpec.describe FailToBan::Strategies::BackoffStrategy do

  subject do
    described_class.new(
      key: key,
      storage: storage
    )
  end

  let(:storage) { NullStorage.new }
  let(:key) { 'dev@jobteaser.com' }

  after { storage.flushdb }

  describe '#attempt' do
    it 'checks against the retry count' do
      allow(Time).to receive(:now).and_return(Time.new('2015', '01', '01').utc)

      3.times { expect(subject.attempt).to eq :ok }
      expect(subject.attempt).to eq :blocked
    end

    it 'does not take old locks into account' do
      storage.hset("fail_to_ban:#{key}", 'retry_count', '5')
      storage.hset(
        "fail_to_ban:#{key}",
        'unlock_at',
        Time.new('2014', '12', '12').utc.to_i.to_s
      )

      allow(Time).to receive(:now).and_return(Time.new('2015', '01', '01').utc)
      expect(subject.attempt).to eq :ok
    end

    it 'takes current locks into account' do
      storage.hset("fail_to_ban:#{key}", 'retry_count', '5')
      storage.hset(
        "fail_to_ban:#{key}",
        'unlock_at',
        Time.new('2014', '12', '12').utc.to_i.to_s
      )

      allow(Time).to receive(:now).and_return(Time.new('2014', '12', '11').utc)
      expect(subject.attempt).to eq :blocked
    end
  end

  describe '#blocked?' do
    it 'is not blocked by default' do
      expect(subject.blocked?).to be(false)
    end

    it 'takes current locks into account' do
      storage.hset("fail_to_ban:#{key}", 'retry_count', '4')
      storage.hset(
        "fail_to_ban:#{key}",
        'unlock_at',
        Time.new('2014', '12', '12').utc.to_i.to_s
      )
      allow(Time).to receive(:now).and_return(Time.new('2014', '12', '11').utc)

      expect(subject.blocked?).to be(true)
    end
  end

  describe '#reset' do
    it 'removes the data associated with the key from storage' do
      storage.hset("fail_to_ban:#{key}", 'retry_count', '0')
      storage.hset("fail_to_ban:#{key}", 'unlock_at', '0')

      expect { subject.reset }.
        to change { storage.hget("fail_to_ban:#{key}", 'retry_count') }.
        from('0').to(nil).
        and change { storage.hget("fail_to_ban:#{key}", 'unlock_at') }.
        from('0').to(nil)
    end
  end

end


lib/fail_to_ban/strategies/backoff_strategy.rb
# frozen_string_literal: true

class FailToBan
  module Strategies
    class BackoffStrategy

      PROTECT_DURATION = 60 * 60 * 24 # eq 24 hours in seconds
      private_constant :PROTECT_DURATION

      JITTER_VARIANCE = 0.2
      private_constant :JITTER_VARIANCE

      HEADER = 'fail_to_ban'
      private_constant :HEADER

      def initialize(key:, storage:, config: {})
        @storage = storage
        @config = default_config.merge(config)
        @id = "#{HEADER}:#{key}"
      end

      def attempt
        return :blocked if blocked?
        increment_failed_attempts
        :ok
      end

      def blocked?
        retry_count >= @config[:permitted_attempts] && Time.now.utc.to_i < unlock_at
      end

      def reset
        @storage.del(@id)
      end

      def unlock_at
        @storage.hget(@id, 'unlock_at').to_i
      end

      private

      def default_config
        {
          permitted_attempts: 3,
          backoff_step: 15, # in seconds
          backoff_cap: nil
        }
      end

      def increment_failed_attempts
        @storage.hincrby(@id, 'retry_count', 1)
        @storage.hset(@id, 'unlock_at', Time.now.utc.to_i + backoff)
        @storage.expire(@id, PROTECT_DURATION)
      end

      def retry_count
        @storage.hget(@id, 'retry_count').to_i
      end

      # Backoff : after the permitted_attempts is exceeded is a x seconds wait
      # If it fails again then it's x * 2 seconds, then x * 3,
      # If the back_time exceeds the backoff cap, then it stop growing
      # In any case, set a +/- 10% jitter on the wait (e.g with a backoff step of 15: 14, 28, 47, ...)
      def backoff
        variance = (1 + JITTER_VARIANCE * (rand - 0.5))
        back_time = @config[:backoff_step] * (retry_count - @config[:permitted_attempts] + 1) * variance
        if !@config[:backoff_cap].nil? && back_time > @config[:backoff_cap] * variance
          @config[:backoff_cap] * variance
        else
          back_time
        end
      end

    end
  end
end
